### ISSUE TITLE
useState setting issue OR Am I using useState horribly wrong?

### ISSUE THREAD

**Issue by DExcalibur** [facebook/react#16075]
**Do you want to request a *feature* or report a *bug*?**
bug  / behavior / or wrong usage by me

**What is the current behavior?**
isFresh remains same as previous state

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**
(https://codesandbox.io/embed/goofy-blackburn-ptcil)
two components
state will be logged in console after every set
click button to set state for component individually
App1 - deepcopy of state object - unexpected behavior
App2 - shallowcopy of state object - works fine

**What is the expected behavior?**
value of isFresh in state should get set to false (in deepcopy part)

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
latest - stable

**Comment by marceloadsj** [comment]
On the **44th** line of **App_deepcopy** you are passing the original **fields** var to **valueSetter**, should be the **fieldsCopy**.

And just my 2¢: the code is really hard to follow, it's using a bool to execute two flows inside the same fn, "deep copying" the array many times. It's okay for the example, but, for me, not so good in a real case.

**Comment by threepointone** [comment]
^ that seems to be a workaround. Consider using a community resource for future questions like this https://reactjs.org/community/support.html
